proxyclient/client.go

```diff
@@ -0,0 +1,239 @@
+	retryTimeout             = 1 * time.Second
+	certificateWatchInterval = 10 * time.Second
+)
+
+type PortForwarder interface {
+	Start() error
+	Stop()
+}
+
+type ProxyClientOpt func(*ProxyClient)
+
+type ProxyClient struct {
+	forwarder           PortForwarder
+	serverUrl           string
+	serverConnectSecret string
+
+	dialer    *websocket.Dialer
+	dialerMtx sync.Mutex
+
+	secretController v1.SecretController
+	namespace        string
+	certSecretName   string
+	certServerName   string
+
+	onConnect func(ctx context.Context, session *remotedialer.Session) error
+}
+
+func New(ctx context.Context, serverSharedSecret, namespace, certSecretName, certServerName string, restConfig *rest.Config, forwarder PortForwarder, opts ...ProxyClientOpt) (*ProxyClient, error) {
+	if restConfig == nil {
+		return nil, fmt.Errorf("restConfig required")
+	}
+
+	if forwarder == nil {
+		return nil, fmt.Errorf("a PortForwarder must be provided")
+	}
+
+	if namespace == "" {
+		return nil, fmt.Errorf("namespace required")
+	}
+
+	if certSecretName == "" {
+		return nil, fmt.Errorf("certSecretName required")
+	}
+
+	if serverSharedSecret == "" {
+		return nil, fmt.Errorf("server shared secret must be provided")
+	}
+
+	serverUrl := fmt.Sprintf("%s:%d%s", defaultServerAddr, defaultServerPort, defaultServerPath)
+
+	client := &ProxyClient{
+		serverUrl:           serverUrl,
+		forwarder:           forwarder,
+		serverConnectSecret: serverSharedSecret,
+		certSecretName:      certSecretName,
+		certServerName:      certServerName,
+		namespace:           namespace,
+	}
+
+	if err := client.buildDialer(ctx, restConfig); err != nil {
+		return nil, fmt.Errorf("dialer build failed %w: ", err)
+	}
+
+	for _, opt := range opts {
+		opt(client)
+	}
+
+	return client, nil
+}
+
+func (c *ProxyClient) buildDialer(ctx context.Context, restConfig *rest.Config) error {
+	core, err := core.NewFactoryFromConfigWithOptions(restConfig, nil)
+	if err != nil {
+		return fmt.Errorf("build secret controller failed: %w", err)
+	}
+
+	secretController := core.Core().V1().Secret()
+	secretController.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(oldSecret, newSecret interface{}) {
+				updatedSecretCert, ok := newSecret.(*corev1.Secret)
+				if ok {
+					if updatedSecretCert.Name == c.certSecretName {
+						rootCAs, err := buildCertFromSecret(c.namespace, c.certSecretName, updatedSecretCert)
+						if err != nil {
+							logrus.Errorf("build certificate failed: %s", err.Error())
+							return
+						}
+
+						c.dialerMtx.Lock()
+						c.dialer = &websocket.Dialer{
+							TLSClientConfig: &tls.Config{
+								RootCAs:    rootCAs,
+								ServerName: c.certServerName,
+							},
+						}
+						c.dialerMtx.Unlock()
+						logrus.Infof("certificate updated successfully")
+					}
+				}
+			},
+		})
+
+	if err := core.Start(ctx, 1); err != nil {
+		return fmt.Errorf("secret controller factory start failed: %w", err)
+	}
```

Comment:
This can be removed since we're just going to pass in the controller, the controller will already be started in Rancher.